en/lessons/basics/testing.md

```diff
@@ -178,12 +178,174 @@ end
+Before we get into more complex use cases, let's talk about some techniques that can help us make our code more testable.  One simple tactic is to pass a module to a function rather than hard-coding the module inside the function.
+
+Instead of code like this, where we have hard-coded the HTTP client used:
+```elixir
+def get_username(username) do
+  HTTPoison.get("https://elixirschool.com/users/#{username}")
+end
+```
+We could instead do something more like this:
+```elixir
+def get_username(username, http_client) do
+  http_client.get("https://elixirschool.com/users/#{username}")
+end
+```
+
+Or perhaps using the [apply/3](https://hexdocs.pm/elixir/Kernel.html#apply/3) function to accomplish the same:
+
+```elixir
+def get_username(username, http_client) do
+  apply(http_client, :get, ["https://elixirschool.com/users/#{username}"])
+end
+```
+
+Passing the module as an argument helps separate concerns and if we don't get too tripped up on the object-oriented verbiage, we might recognize this inversion of control as a kind of [Dependency Injection](https://en.wikipedia.org/wiki/Dependency_injection).  To test the `get_username/2` method, you would only need to pass in a module whose `get` function returned the value needed for your assertions.
+
+This construct is very simple, so it is only useful when the function is highly accessible (and not, for example, buried somewhere deep within a private function).
+
+A more flexible tactic relies on the application configuration. Perhaps you didn't even realize it, but an Elixir application maintains state in its configuration. Rather than hard-coding a module or passing it as an argument, you can read it from the application config.
+
+```elixir
+def get_username(username) do
+  http_client().get("https://elixirschool.com/users/#{username}")
+end
+
+defp http_client do
+  Application.get_env(:my_app, :http_client)
+end
+```
+
+Then, in your config file:
+
+```elixir
+config :my_app, :http_client, HTTPoison
+```
+
+This construct and its reliance on the application config forms the basis of everything that follows.
+
+If you are prone to overthinking, yes, you could omit the `http_client/0` function and call `Application.get_env/2` directly, and yes, you could also provide a default third argument to `Application.get_env/3` and achieve the same result.
+
+This technique allows us to have specific implementations of the module for each environment; in José's article, he referenced a sandbox module for the `dev` environment while the `test` environment used an in-memory module.
+
+However, having only one fixed module per environment may not be flexible enough: depending on how your function is used, you may need to return different responses in order to test all possible execution paths. What most people don't know is that you can _change_ the application configuration at run-time! Let's take a look at [Application.put_env/3](https://hexdocs.pm/elixir/Application.html#put_env/4).
```

Comment:
Is there any reason to stick with `Application.put_env/3`, not put_env/4?

Reply:
Good catch.  Both `put_env/3` and `put_env/4` are in the docs and link to the same function: 
- https://hexdocs.pm/elixir/Application.html#put_env/3
- https://hexdocs.pm/elixir/Application.html#put_env/4

So... to be honest, I don't know which one is technically correct when it comes to the the default value...

Reply:
I think that `put_env/4` is correct and can be used as `put_env/3` with default values.
The table of contents also displays as `put_env/4`.

<img width="220" alt="Screen Shot 2020-06-23 at 11 25 52" src="https://user-images.githubusercontent.com/128431/85353966-67acc180-b544-11ea-9faf-3677c6c2ba4d.png">

Reply:
No. We can start going though and updating the lessons to leverage the newer APIs :+1: